### PR TITLE
Instanced 3D Model spec

### LINKED ISSUE #23: Instanced 3D Model and Composite Tile Formats
First, see the original notes for [Instanced 3D Model](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/TileFormats/Instanced3DModel/README.md) and [Composite](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/TileFormats/Composite/README.md) tile formats. To help design the Instanced 3D Model tile format, see the [Batched 3D Model](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/b3dm/README.md) tile format.

We now have significant work towards instancing in Cesium.  See https://github.com/AnalyticalGraphicsInc/cesium/pull/3049.

Here's some initial thoughts:

## Instanced 3D Model

```
Header
------
magic            // uchar[4], "i3dm"
version          // uint32,   1
batchTableLength // uint32,   can be zero, length of the batch table in bytes
glTFLength       // uint32,   length of glTF section (after the Batch Table) in bytes
glTFFormat       // uint32,   0 = url, 1 = embedded binary glTF
instancesLength  // uint32,   number of instances, NOT number of bytes, can be zero I guess

Batch Table
-----------
// Same as Batched 3D Model.  Optional.

glTF
----
// UTF-8 string or binary blob depending on Header.glTFFormat
// Length in bytes is Header.glTFLength

Instances
---------
There are Header.instancesLength of these:
* x, y, z    // doubles in WGS84 coordinates
* x, y, z, w // doubles in WGS84 coordinates (quaternion)
* batchId    // uint32, index into batchTable, zero when Header.batchTableLength === 0
```

* [x] ~~We can significantly optimize the size of each `instance` field, e.g., positions can be quantized relative to the bounding volume (see [EXT_quantized_attributes](https://github.com/mlimper/glTF/tree/spec-1.0/extensions/MultiVendor/EXT_quantized_attributes)).  Rotation could also be optional (part of the batchTable) and default to Y up (glTF default) in a local frame for the instance's position.  Rotation could use fewer fields.  RTC.   Delta encoding... SOA vs. AOS...  Look into point cloud compression...~~ #33
* [x] ~~Should `instance.batchId` be omitted when `header.batchTableLength === 0`.  This will complicate the client, but save memory; however, the case it saves memory for is probably rare in practice.~~ #33
* [x] ~~Throughout 3D Tiles, should `batchTable` be renamed?  Something like `propertyTable`, `metadataTable`, etc.  We can also decide this later as we define more tile formats.  Consistent terminology is important.~~ Moved to #32
* [x] ~~Throughout 3D Tiles, should the `batchTable` work more like glTF buffer/bufferView/accessor to avoid arrays of numbers?~~ Moved to #32

## Composite

In order to have different models instanced in the same tile (or more generally, combine any number of tiles, even tiles of different formats), we can use a Composite tile format.

Perhaps:
```
Header
------
magic            // uchar[4], "cmpt"
version          // uint32,   1
tilesLength      // Number of tiles

Tiles
----
// header.tilesLength (potentially heterogeneous) tiles
```

* [x] To make this easier to parse on the client, each tile's header could include its total length in bytes or the Composite can contain an array of lengths.  I think we want to store it in each tile so other binary streams containing tiles can be easily parsed.

### PR DESCRIPTION
**Pull request by pjcozzi**
Fixes #23

@slchow please copyedit.  The new writing is in `TileFormats/Instanced3DModel/README.md`.  Note that most of this PR is the same changes as #31.

@lilleyse please do a technical review.  Does the Cesium client handle the corner cases for when, for example, `instancesLength` or `batchTableLength` is zero?

@lilleyse merge #31 first.

Before we declare 1.0, there's a few things to address in #33

### CODE REVIEW

**Comment by lilleyse** [comment]
* Right now I'm using `uint16` for `batchId`, but we could go with `uint32` as well.
* Should `gltfLength` and `batchTableLength` be rename with `byteLength`?
* I tested the corner cases, and they work correctly.

File: TileFormats/Instanced3DModel/README.md [hunk 1]
```diff
@@ -1,13 +1,113 @@
...
+
+* a binary blob containing binary glTF.
+
+In either case, `header.glTFLength` contains the length of the glTF field in bytes.
+
+## Instances
+
+The `instances` field immediately follows the `glTF` field (which may be omitted when `header.glTFLength` is `0`).
+
+The `instances` field contains `header.instancesLength` tightly packed instances.  Each instance has three fields:
```
Comment (slchow) [inline TileFormats/Instanced3DModel/README.md]: Is there a word missing here? I couldn't make sense of "contains `header.instancesLength` tightly packed instances"

Comment (pjcozzi) [inline TileFormats/Instanced3DModel/README.md]: A word is not missing.  That should read as "the instances field contains x tightly packed instances" where the value of "x" is defined by `header.instancesLength`.

If you think we could make it more clear, please have at it.

**Comment by pjcozzi** [comment]
> Right now I'm using uint16 for batchId, but we could go with uint32 as well.

`uint16` is OK.  That allows 64K per tile.  That ought to be enough for anybody. :)

If experience proves otherwise, we can make this `unit16` or `uint32` depending on the value of `header.batchTableLength` or just `uint32` if compression make the size difference not matter.

> Should gltfLength and batchTableLength be rename with byteLength?

Yes, I will make the rename for b3dm and i3dm in this branch.

**Comment by pjcozzi** [comment]
> If experience proves otherwise, we can make this unit16 or uint32 depending on the value of header.batchTableLength or just uint32 if compression make the size difference not matter.

Sorry, I mean the length of the arrays in the batch table, not the size of the JSON string in bytes.

**Comment by pjcozzi** [comment]
> Should gltfLength and batchTableLength be rename with byteLength?

Also, please make these same renames throughout the Cesium implementation.

`glTFLength` -> `gltfByteLength`

`glTFFormat` -> `gltfFormat`

`batchTableLength` -> `batchTableByteLength`

**Comment by pjcozzi** [comment]
@lilleyse this is updated.  I'll do the b3dm `batchTableByteLength` rename in the `composite` branch, which has the other b3dm changes, now.

@slchow are you still copyediting?

**Comment by slchow** [comment]
@pjcozzi I'm done.

**Comment by pjcozzi** [comment]
Thanks @slchow!

@lilleyse merge when ready.

**Comment by lilleyse** [comment]
The layout image still has `batchId` as `uint32`, Other than that, it looks good to me.

**Comment by pjcozzi** [comment]
Good catch, @lilleyse.  Updated.

**Comment by lilleyse** [comment]
Merged.

**Pull request merged**